src/Libraries/Microsoft.Extensions.DataIngestion/Writers/VectorStoreWriter.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.VectorData;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.DataIngestion;
+
+/// <summary>
+/// Writes chunks to the <see cref="VectorStore"/> using dynamic schema. 
+/// </summary>
+/// <typeparam name="T">The type of the chunk content.</typeparam>
+public sealed class VectorStoreWriter<T> : IngestionChunkWriter<T>
+{
+    // The names are lowercase with no special characters to ensure compatibility with various vector stores.
+    private const string KeyName = "key";
+    private const string EmbeddingName = "embedding";
+    private const string ContentName = "content";
+    private const string ContextName = "context";
+    private const string DocumentIdName = "documentid";
+
+    private readonly VectorStore _vectorStore;
+    private readonly int _dimensionCount;
+    private readonly VectorStoreWriterOptions _options;
+    private readonly bool _keysAreStrings;
+
+    private VectorStoreCollection<object, Dictionary<string, object?>>? _vectorStoreCollection;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="VectorStoreWriter{T}"/> class.
+    /// </summary>
+    /// <param name="vectorStore">The <see cref="VectorStore"/> to use to store the <see cref="IngestionChunk{T}"/> instances.</param>
+    /// <param name="dimensionCount">The number of dimensions that the vector has. This value is required when creating collections.</param>
+    /// <param name="options">The options for the vector store writer.</param>
+    /// <exception cref="ArgumentNullException">When <paramref name="vectorStore"/> is null.</exception>
+    /// <exception cref="ArgumentOutOfRangeException">When <paramref name="dimensionCount"/> is less or equal zero.</exception>
+    public VectorStoreWriter(VectorStore vectorStore, int dimensionCount, VectorStoreWriterOptions? options = default)
```

Comment:
Should there be a constructor that directly accepts the VectorStoreCollection? That would allow users to instantiate a collection with whatever database-specific config they might need, and then wrap that for ingestion.

Reply:
We could add that, but I am afraid that this would also require the customers to provide a mapper (as soon as they use a different name/type we are doomed). Having said that, I am open to add that after first preview.